semgrep-core/parsing/Parse_tree_sitter_helpers.ml

```diff
@@ -92,6 +92,45 @@ let combine_tokens env xs =
       let t = token env x in
       t
 
+let mk_tree_sitter_error (err : Tree_sitter_run.Tree_sitter_error.t) =
+  let start = err.start_pos in
+  let loc = {
+    PI.str = err.substring;
+    charpos = 0; (* fake *)
+    line = start.row + 1;
+    column = start.column;
+    file = err.file.name;
+  } in
+  loc
+
+let convert_tree_sitter_exn_to_pfff_exn f =
+  try f ()
+  with
+  (* The case below is what we would like to do! However if
+   * you use Parallel.invoke to invoke the tree-sitter parser, this
+   * code below will never trigger. Indeed, unmarshalled exn
+   * can't be used in match or try or used for structural equality
+   * hence the ugly workaround below. See marshal.mli or Paralle.ml for
+   * more information.
+   *)
+  | Tree_sitter_run.Tree_sitter_error.Error ts_error ->
+    let loc = mk_tree_sitter_error ts_error in
+    let info = { PI.token = PI.OriginTok loc; transfo = PI.NoTransfo } in
+    raise (PI.Parsing_error info)
+
+  (* !!!UGLY!!! remove this once we don't use Paralle.invoke *)
+  | exn ->
+      let s = Common.exn_to_s exn in
+      if s = "Tree_sitter_run.Tree_sitter_error.Error(_)" then begin
+        let t = Obj.repr exn in
+        let info = Obj.field t 1 in
+        let (ts_error : Tree_sitter_run.Tree_sitter_error.t) = Obj.obj info in
+        let loc = mk_tree_sitter_error ts_error in
+        let info = { PI.token = PI.OriginTok loc; transfo = PI.NoTransfo } in
+        raise (PI.Parsing_error info)
+      end else raise exn
```

Comment:
To get the constructor, you can do `Printexc.exn_slot_name exn` (since ocaml 4.02). But I don't see how to do the rest without `Obj`, other than catching the exception in the child (or make the parser return an Ok/Error).